view/log.go

```diff
@@ -149,6 +149,136 @@ func (li *logView) Read(facts origins.Facts) (int, error) {
 	return l, nil
 }
 
+// merger maintains state of a merged stream of facts from multiple domains.
+type merger struct {
+	iterators []origins.Iterator
+	next      []*origins.Fact
+}
+
+func (m *merger) Next() (*origins.Fact, error) {
+	var maxTrn uint64 = 0
+	var maxI int = -1
+
+	// look at the next fact coming out of each of the the available logs,
+	// and choose the fact with the greatest transaction ID.
+	for i := 0; i < len(m.iterators); i++ {
+		if m.next[i] != nil && m.next[i].Transaction > maxTrn {
+			maxTrn = m.next[i].Transaction
+			maxI = i
+		}
+	}
+
+	if maxI == -1 {
+		return nil, io.EOF
+	}
+
+	// push the newest fact out of the merged stream
+	nextFact := m.next[maxI]
+	fact, err := m.iterators[maxI].Next()
+
+	if err != nil {
+		if err == io.EOF {
+			m.next[maxI] = nil
+		} else {
+			return nil, err
+		}
+	} else {
+		m.next[maxI] = fact
+	}
+
+	return nextFact, nil
+}
+
+func (m *merger) init() error {
```

Comment:
Since this is only called once, the initialization code should go in to the `Merge` function below.